addons/website_sale_loyalty/models/sale_order.py

```diff
@@ -156,7 +156,18 @@ def get_promo_code_success_message(self, delete=True):
             request.session.pop('successful_code')
         return code
 
+    def _is_to_delete_discount(self, line, set_qty):
+        self.ensure_one()
+        return set_qty == 0 and line.coupon_id and line.reward_id and line.reward_id.reward_type == 'discount'
+
     def _cart_update(self, *args, **kwargs):
+        product_id, set_qty = kwargs['product_id'], kwargs.get('set_qty')
+
+        # Check if it's required to delete the line of the discount code
+        line = self.order_line.filtered(lambda l: l.product_id.id == product_id)
+        if set_qty == 0 and line.coupon_id and line.reward_id and line.reward_id.reward_type == 'discount':
+            kwargs['line_id'] = line.id
```

Comment:
```suggestion
        line = self.order_line.filtered(lambda l: l.product_id.id == product_id)
        if set_qty == 0 and line.coupon_id and line.reward_id and line.reward_id.reward_type == 'discount':
            # Force the deletion of the line even if it's a temporary record created by new()
            kwargs['line_id'] = line.id
```
I changed the comment and put it in the if, I think it's clearer like this